src/compiler/tools/LanguageService.ts

```diff
@@ -11,6 +11,18 @@ import { Program } from "./Program";
 import { DefinitionInfo, EmitOutput, FileTextChanges, ImplementationLocation, RenameLocation, TextChange, DiagnosticWithLocation, RefactorEditInfo, CodeFixAction,
     CombinedCodeActions } from "./results";
 
+export interface CompilerHostOverrides {
+    resolveModuleNames?(moduleNames: string[], containingFile: string, reusedNames?: string[], ref?: ts.ResolvedProjectReference): (ts.ResolvedModule | undefined)[];
+    getResolvedModuleWithFailedLookupLocationsFromCache?(modulename: string, containingFile: string): ts.ResolvedModuleWithFailedLookupLocations | undefined;
+    resolveTypeReferenceDirectives?(typeDirectiveNames: string[], containingFile: string, ref?: ts.ResolvedProjectReference): (ts.ResolvedTypeReferenceDirective | undefined)[];
+}
+
+export type CompilerHostFactory = (languageService: ts.LanguageServiceHost) => CompilerHostOverrides;
```

Comment:
Thanks! Will look at this PR later. Question though: why does it require a language service host? Could you provide an example of some code that uses this?

Reply:
Sure, everything is pretty clear. First of all,  `LanguageService` implements almost everything that is the need for work with sources and file system, so it is really useful for module resolution.

And another thing is `resolveTypeReferenceDirectives` implementation.  The default implementation is referenced from `ts.resolveTypeReferenceDirective` function, that is exposed directly on `ts` namespace. This function has the following signature: 
```
function resolveTypeReferenceDirective(typeReferenceDirectiveName: string, containingFile: string | undefined, options: CompilerOptions, host: ModuleResolutionHost, redirectedReference?: ResolvedProjectReference): ResolvedTypeReferenceDirectiveWithFailedLookupLocations;
```
As you can see, `ModuleResolutionHost` is required argument that can be build only with 'LanguageService' implemenatations. Here is `ModuleResolutionHost` interface:
```
 interface ModuleResolutionHost {
        fileExists(fileName: string): boolean;
        readFile(fileName: string): string | undefined;
        trace?(s: string): void;
        directoryExists?(directoryName: string): boolean;    
        realpath?(path: string): string;
        getCurrentDirectory?(): string;
        getDirectories?(path: string): string[];
    }
```
Everything about FS. 
Such a strategy is used in TypeScript - (source)[https://github.com/microsoft/TypeScript/blob/master/src/harness/harnessLanguageService.ts#L295] where `NativeLanguageServiceHost` is an implementation of `ts.LanguageServiceHost`

Reply:
So it seems like the main thing that's absolutely necessary is the ability to have access to the project's file system?

I'm thinking that this method should instead be passed a `FileSystemHost` and then someone could construct a `ts.ModuleResolutionHost` from that if they need it (and perhaps a helper function could be added so people could easily create a module resolution host from a `FileSystemHost`). I'd rather not expose the internally used `ts.CompilerHost` because then it becomes part of the public api.

Reply:
For example, we can inject `FileSystemHost` from `Project`.